activesupport/lib/active_support/environment_inquirer.rb

```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  # This is a special case of StringInquirer that defines the three default
+  # environments at construction time based on the environment string.
+  class EnvironmentInquirer < StringInquirer
+    DEFAULT_ENVIRONMENTS = ["development", "test", "production"]
+    def initialize(env)
+      super(env)
+
+      DEFAULT_ENVIRONMENTS.each do |default_env|
+        singleton_class.define_method(:"#{env}?", (env == default_env).method(:itself))
+      end
```

Comment:
Do we need to do this in initialize on the singleton class?  Why not just make regular methods that do this work?  I understand that this optimizes out the comparison on subsequent calls, but is that really a necessary optimization?

Reply:
There's three options as I see it:

* This one, which is the fastest (lowest overhead) per call to the query methods. Main cost is having the singleton.
* Storing ivars for `@development` etc and calculating them at init time. Additional cost is retrieving the var and a subclass of String that adds ivars.
* Defining methods that compare self to `"development"` etc. Additional cost is string comparison on each call.

I'd like to avoid introducing block-based `define_method` methods since they still have more overhead than all other types of methods on CRuby. The second option above is probably acceptable, but is only like 6x faster than the method_missing 🤷‍♂️

Reply:
Consider letting the base class take responsibility for hinted optimizations and keeping env inquiry concerns in Railties:

```ruby
class StringInquirer
  def initialize(string, inquiry_hints: [ … ])
    …
    define_hinted_inquiries(inquiry_hints)
  end

  private
    def define_hinted_inquiries(hints)
      hints.each { … }
    end
end
```

Then we could do a private class in Railties:

```ruby
class RailsEnvironmentInquirer < StringInquirer
  def initialize(environment)
    super environment,  %i[ development test production ]
  end
end
private_constant :RailsEnvironmentInquirer

def env
  @_env ||= RailsEnvironmentInquirier.new(…)
```

Could dispense with a subclass altogether, but that's a bit fussier:
```ruby
DEFAULT_ENVIRONMENTS = %i[ development test production ]

def env=(environment)
  @_env = ActiveSupport::StringInquirer.new(environment, DEFAULT_ENVIRONMENTS)
end

# changing this up so we don't repeat StringInquirer.new invocation
def env
  unless defined?(@_env)
    self.env = ENV["RAILS_ENV"].presence || ENV["RACK_ENV"].presence || "development"
  end
  @_env
end
```

Reply:
I'd rather not add functionality to the public base class. This is just about fixing one common usage, not making this a customizable query method delegate. That's too much scope creep here.

I would however be interested in making the env class private and perhaps passing the environments in from a location more suited to knowing them.